### ISSUE TITLE
Add walk-forward evaluation engine

### ISSUE THREAD

**Issue by 5queezer** [5queezer/basana#22]
<h2>Summary</h2>
<p>The <code>basana.core.evaluation</code> module has window generators (<code>generate_sliding_windows</code>, <code>generate_expanding_windows</code>), data types (<code>FoldResult</code>, <code>EvaluationReport</code>), and reporting (<code>build_report</code>, <code>format_report</code>) — but no engine to actually <strong>run</strong> a strategy across folds. Users currently have to write their own loop to execute backtests per window, collect trades/equity, and assemble the report.</p>
<p>This proposes adding <code>WalkForwardEngine</code> to close that gap.</p>
<h2>Motivation</h2>
<p>Walk-forward evaluation is the standard way to validate trading strategies without lookahead bias. The existing pieces handle "what windows to create" and "how to format results," but the critical middle step — "run a strategy factory across train/test pairs and collect metrics" — is left to the user. This means every strategy author re-implements the same fold loop, metric collection, and OOS aggregation logic.</p>
<h2>Proposed API</h2>
<pre><code class="language-python">from basana.core.evaluation import WalkForwardEngine, FoldOutput, generate_sliding_windows, format_report

# User provides an async callable that runs one backtest for a WindowSpec
async def my_runner(window: WindowSpec, params: dict | None) -&gt; FoldOutput:
    # set up dispatcher, exchange, strategy for window.start → window.end
    # run backtest
    # return FoldOutput(trades=ledger.trades, equity_curve=ledger.equity_curve)
    ...

# Optional: optimizer extracts best params from training output
async def my_optimizer(window: WindowSpec, train_output: FoldOutput) -&gt; dict:
    # e.g. sweep hyperparams on training data, return best
    return {"kp": 1.2, "ki": 0.1, "kd": 3.6}

engine = WalkForwardEngine(
    strategy_name="pid_mean_reversion",
    strategy_runner=my_runner,
    windows=generate_sliding_windows(start, end, train_duration, test_duration),
    param_optimizer=my_optimizer,       # optional
    periods_per_year=365 * 24,
)
report = await engine.run()
print(format_report(report))
</code></pre>
<h2>Design</h2>
<p><strong><code>WalkForwardEngine</code></strong> orchestrates the loop:</p>
<ol>
<li>Pairs windows into (train, test) tuples via existing <code>pair_windows</code>.</li>
<li>For each fold:
<ul>
<li>Runs <code>strategy_runner(train_window, None)</code> → gets <code>FoldOutput</code>.</li>
<li>If <code>param_optimizer</code> is set, calls it with the training output → gets params dict.</li>
<li>Runs <code>strategy_runner(test_window, params)</code> → gets test <code>FoldOutput</code>.</li>
<li>Builds a <code>FoldResult</code> with train/test metrics.</li>
</ul>
</li>
<li>Concatenates all OOS trades and equity snapshots (sorted chronologically).</li>
<li>Calls existing <code>build_report</code> to produce the <code>EvaluationReport</code>.</li>
</ol>
<p><strong><code>FoldOutput</code></strong> is a simple container returned by the strategy runner:</p>
<ul>
<li><code>trades: Sequence[TradeRecord]</code></li>
<li><code>equity_curve: Sequence[EquitySnapshot]</code></li>
<li><code>metrics: Optional[PerformanceMetrics]</code> — pre-calculated, or <code>None</code> to auto-calculate via <code>calculate_metrics</code></li>
<li><code>params: Optional[Dict[str, Any]]</code> — records what params were used</li>
</ul>
<p><strong><code>StrategyRunner</code></strong> and <strong><code>ParamOptimizer</code></strong> are type aliases for the async callables, making the expected signatures explicit.</p>
<h3>What it doesn't do</h3>
<ul>
<li>No parallel execution (folds run sequentially). Can be added later.</li>
<li>No CSV/data filtering by window dates — the runner is responsible for setting up the right bar source for the given <code>WindowSpec</code>. This keeps the engine exchange-agnostic.</li>
<li>No built-in param sweep — the <code>param_optimizer</code> is a user-provided callable, so any optimization approach (grid search, Bayesian, random) works.</li>
</ul>
<h2>Files</h2>

File | Change
-- | --
basana/core/evaluation/engine.py | New — WalkForwardEngine, FoldOutput, type aliases
basana/core/evaluation/__init__.py | Add exports for WalkForwardEngine, FoldOutput, StrategyRunner, ParamOptimizer
tests/test_walk_forward_engine.py | New — 8 tests

[engine.py](https://github.com/user-attachments/files/25993177/engine.py)

[__init__.py](https://github.com/user-attachments/files/25993179/__init__.py)

[test_walk_forward_engine.py](https://github.com/user-attachments/files/25993181/test_walk_forward_engine.py)

<h2>Tests</h2>
<ul>
<li><code>test_engine_basic</code> — runs folds, verifies runner called for train+test, aggregate OOS collected</li>
<li><code>test_engine_with_optimizer</code> — params from optimizer forwarded to test runner</li>
<li><code>test_engine_empty_windows</code> — graceful no-op on empty input</li>
<li><code>test_engine_fold_metrics</code> — per-fold train/test metrics populated correctly</li>
<li><code>test_engine_pre_calculated_metrics</code> — respects <code>FoldOutput.metrics</code> when provided</li>
<li><code>test_engine_report_is_formattable</code> — output works with existing <code>format_report</code></li>
<li><code>test_engine_oos_equity_sorted</code> — aggregate equity curve sorted chronologically</li>
<li><code>test_fold_output_params</code> — <code>FoldOutput</code> stores optional params</li>
</ul>
<p>All 8 new tests pass. Existing 16 evaluation tests unaffected.</p></body></html>
